manager/manager.go

```diff
@@ -1086,22 +1086,26 @@ func (m *manager) destroyContainerLocked(containerName string) error {
 
 // Detect all containers that have been added or deleted from the specified container.
 func (m *manager) getContainersDiff(containerName string) (added []info.ContainerReference, removed []info.ContainerReference, err error) {
-	m.containersLock.RLock()
-	defer m.containersLock.RUnlock()
-
 	// Get all subcontainers recursively.
+	m.containersLock.RLock()
 	cont, ok := m.containers[namespacedContainerName{
 		Name: containerName,
 	}]
+	m.containersLock.RUnlock()
 	if !ok {
 		return nil, nil, fmt.Errorf("failed to find container %q while checking for new containers", containerName)
 	}
-	allContainers, err := cont.handler.ListContainers(container.ListRecursive)
+	contHandler := cont.handler
```

Comment:
I think, it's better leave assignation, but move it under the lock, to avoid possible data race.

Reply:
ok

Reply:
With assignation placed under the lock, code looks ugly without `defer`, but with `defer` we returning to the first variant. Maybe, it's better to use initial variant?

```
	m.containersLock.RLock()
	cont, ok := m.containers[namespacedContainerName{
		Name: containerName,
	}]
	if !ok {
		m.containersLock.RUnlock()
		return nil, nil, fmt.Errorf("failed to find container %q while checking for new containers", containerName)
	}
	contHandler := cont.handler
	m.containersLock.RUnlock()
```

Reply:
I am not convinced there is a data race in reading the handler attribute of `cont`.  It is not modified once it is initialized.  The containersLock is meant to prevent concurrent access to the `containers` map, which the  current design does.  I don't think you need to add a new `contHandler` variable here.

Reply:
Yes, there are no data races in the current codebase, but i have one concern:
When m.containersLock locked, we expect exclusive access to the m.containers and it's members' fields. It's unexpected, that cont.handler could be read by anyone else during lock, so it could lead to tricky data race in future.
What do you think about it?